src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,483 @@
+         * @throws NullPointerException
+         *     if the {@code ikm} is null
+         */
+        public Builder addIKM(SecretKey ikm) {
+            Objects.requireNonNull(ikm, "ikm must not be null");
+            ikms.add(ikm);
+            return this;
+        }
+
+        /**
+         * Adds input key material to the builder.
+         * <p>
+         * {@code addIKM} may be called when the input key material value is to
+         * be assembled piece-meal or if part of the IKM is to be supplied by a
+         * hardware crypto device. This method appends to the existing list of
+         * values or creates a new list if there is none yet.
+         * <p>
+         * This supports the use-case where a label can be applied to the IKM
+         * but the actual value of the IKM is not yet available.
+         * <p>
+         * An implementation should concatenate the input key materials into a
+         * single value once all components are available.
+         *
+         * @param ikm
+         *     the input key material value
+         *
+         * @return this builder
+         *
+         * @throws NullPointerException
+         *     if the {@code ikm} is null
+         */
+        public Builder addIKM(byte[] ikm) {
+            Objects.requireNonNull(ikm, "ikm must not be null");
+            if (ikm.length != 0) {
+                return addIKM(new SecretKeySpec(ikm, "Generic"));
+            } else {
+                return this;
+            }
+        }
+
+        /**
+         * Adds a salt to the builder.
+         * <p>
+         * {@code addSalt} may be called when the salt value is to be assembled
+         * piece-meal or if part of the salt is to be supplied by a hardware
+         * crypto device. This method appends to the existing list of values or
+         * creates a new list if there is none yet.
+         * <p>
+         * An implementation should concatenate the salt into a single value
+         * once all components are available.
+         *
+         * @param salt
+         *     the salt value
+         *
+         * @return this builder
+         *
+         * @throws NullPointerException
+         *     if the {@code salt} is null
+         */
+        public Builder addSalt(SecretKey salt) {
+            Objects.requireNonNull(salt, "salt must not be null");
+            salts.add(salt);
+            return this;
+        }
+
+        /**
+         * Adds a salt to the builder.
+         * <p>
+         * {@code addSalt} may be called when the salt value is to be assembled
+         * piece-meal or if part of the salt is to be supplied by a hardware
+         * crypto device. This method appends to the existing list of values or
+         * creates a new list if there is none yet.
+         * <p>
+         * An implementation should concatenate the salt into a single value
+         * once all components are available.
+         *
+         * @param salt
+         *     the salt value
```

Comment:
It's not clear that these are wrapped in`SecretKey` objects and returned in the `salts()` method of `Extract` and `ExtractThenExpand` - I think that detail needs to be noted in the API, otherwise users might find it surprising. Same comment for `addIKM(byte[])`.